pkg/apis/configuration/v1alpha1/kongcustomentity_types.go

```diff
@@ -0,0 +1,76 @@
+package v1alpha1
+
+import (
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+
+// KongCustomEntityList is a list of kong custom entities.
+type KongCustomEntityList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []KongCustomEntity `json:"items"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=kongcustomentities,shortName=kce,categories=kong-ingress-controller
+
+// KongCustomEntity represents a custom entity in Kong.
+type KongCustomEntity struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// Type is the type of this custom entity.
+	// Should be same as the `Name` of a KongCustomEntityDefinition.
+	Type string `json:"type"`
+	// Fields is the list of fields in the entity.
+	Fields apiextensionsv1.JSON `json:"fields"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster,path=kongcustomentitydefinitions,shortName=kced,categories=kong-ingress-controller
+
+// KongCustomEntityDefinition represents definition of a custom entity type in Kong.
+type KongCustomEntityDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// Name is the type name of the entity.
+	Name string `json:"name"`
```

Comment:
Maybe EntityName for this? I feel like we should distinguish this from the Kubernetes resource name, even if they're in separate parts of the object. Otherwise docs could be awkward, since you'd reference `Name` without it being clear which you're referring to.

This and the other fields should call out the DAO definition field names (`name`, `admin_api_name`, `admin_api_nested_name`) explicitly with "this value should match the `name` field in the Lua entity definition" or similar--even if they're the same names, that relationship may not be obvious.

It's probably still not obvious for entities that first-party plugins provide (since users aren't familiar with the source), but I don't know of a great way to handle that other than calling it out in the respective plugin docs pages.

Reply:
Moved the fields to `spec` so this comes to `spec.name`. If we want to give it a field other than `Name`, I think `TypeName` or `Type` would be better because this is actually the name of a entity type, not the name of an entity.